### DISCUSSION TITLE
High level interface vs low level interface for qp

### DISCUSSION THREAD

**Discussion by junhyeokahn** [casadi/casadi#3357] [Q&A] [answered]
Hello,
Based on the documentation, it looks like there are two ways to formulte qp problem. (qpsol vs conic). Is there any performance difference between these two?
When to use one over the other?

**Comment by jaeandersson** [comment, marked as answer]
The high-level interface (qpsol) is just a wrapper around the low-level interface (conic). The qpsol gives an interface which is equivalent to that of the NLP solver (nlpsol), so you can easily switch between QP solver and an NLP solver. The structure of the QP is automatically detected using CasADi's AD framework. The low-level interface makes sense if you have a problem in QP standard form 

$\text{minimize} \frac{1}{2} x^T Q x +c^T x \quad \text{subject to} \quad x_{lb} \le x \le x_{ub}, \quad a_{lb} \le A  x \le a_{ub}$

In particular, it gives you better control of what is actually sent to the solver. In general, I don't think there will be a major performance difference, since most of the overhead is during the initialization. During the actual solving, there should be minimal overhead. An exception might be if matrices A and Q above are very large and dense.

**Reply by junhyeokahn** [reply]
Thanks for the detailed answer!
